torch/csrc/jit/mobile/nnc/context.cpp

```diff
@@ -0,0 +1,252 @@
+
+namespace {
+
+c10::IValue Tup(std::vector<c10::IValue>&& ivalues) {
+  return c10::ivalue::Tuple::create(ivalues);
+}
+
+} // namespace
+
+InputSpec::InputSpec(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  sizes_ = dict.at("sizes").toIntVector();
+  dtype_ = dict.at("dtype").toScalarType();
+}
+
+c10::IValue InputSpec::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("sizes", sizes_);
+  dict.insert("dtype", dtype_);
+  return dict;
+}
+
+bool InputSpec::validate(const at::Tensor& input) const {
+  return input.sizes() == sizes_ && input.scalar_type() == dtype_;
+}
+
+OutputSpec::OutputSpec(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  sizes_ = dict.at("sizes").toIntVector();
+  dtype_ = dict.at("dtype").toScalarType();
+}
+
+c10::IValue OutputSpec::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("sizes", sizes_);
+  dict.insert("dtype", dtype_);
+  return dict;
+}
+
+at::Tensor OutputSpec::allocate() const {
+  return at::empty(
+      sizes_,
+      at::TensorOptions()
+          .dtype(dtype_)
+          .layout(at::kStrided)
+          .device(at::kCPU)
+          .requires_grad(false));
+}
+
+MemoryPlan::MemoryPlan(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  buffer_sizes_ = dict.at("buffer_sizes").toIntVector();
+}
+
+c10::IValue MemoryPlan::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+  dict.insert("buffer_sizes", buffer_sizes_);
+  return dict;
+}
+
+void MemoryPlan::allocate(ExecutionState* state) const {
+  auto& allocations = state->preallocations_;
+  allocations.reserve(buffer_sizes_.size());
+  for (int64_t buffer_size : buffer_sizes_) {
+    at::DataPtr buffer = c10::GetCPUAllocator()->allocate(buffer_size);
+    allocations.emplace_back(std::move(buffer));
+  }
+}
+
+Function::Function(const c10::IValue& value) {
+  auto dict = value.toGenericDict();
+  name_ = c10::QualifiedName(dict.at("name").toStringRef());
+  nnc_kernel_id_ = dict.at("nnc_kernel_id").toStringRef();
+  parameters_ = dict.at("parameters").toTensorVector();
+
+  // input_specs_
+  for (const auto& input_value : dict.at("input_specs").toTuple()->elements()) {
+    input_specs_.emplace_back(input_value);
+  }
+
+  // output_specs_
+  for (const auto& output_value :
+       dict.at("output_specs").toTuple()->elements()) {
+    output_specs_.emplace_back(output_value);
+  }
+
+  // memory_plan_
+  memory_plan_ = MemoryPlan(dict.at("memory_plan"));
+}
+
+c10::IValue Function::serialize() const {
+  c10::Dict<c10::IValue, c10::IValue> dict(
+      at::StringType::get(), at::AnyType::get());
+
+  dict.insert("name", name_.qualifiedName());
+  dict.insert("nnc_kernel_id", nnc_kernel_id_);
+  dict.insert("parameters", parameters_);
```

Comment:
Interesting that parameters are associated with function and not module. Why not have equivalent module that owns the parametser?

Reply:
Good question - I'll add a TODO here. I'll improve it when we have more than one methods to compile.
I think parameters should not only be persisted with module (so multiple functions can share) but also be shared between the compiled model and the original model.
I mentioned in the design doc that eventually the compiled model should be an addition to the original model (like lite-interpreter's bytecode is an addition to the original model). We just recently solved the parameter sharing problem for lite-interpreter, so it will take some effort.

Reply:
> but also be shared between the compiled model and the original model.

Exactly right. I think this may require some change to backend interface for it to allow querying parameters in some "universal" space.